Lib/pdb.py

```diff
@@ -438,7 +439,30 @@ def default(self, line):
         locals = self.curframe_locals
         globals = self.curframe.f_globals
         try:
-            code = compile(line + '\n', '<stdin>', 'single')
+            if (code := codeop.compile_command(line + '\n', '<stdin>', 'single')) is None:
+                # Multi-line mode
+                buffer = line
+                continue_prompt = "...   "
+                while (code := codeop.compile_command(buffer, '<stdin>', 'single')) is None:
+                    if self.use_rawinput:
+                        try:
+                            line = input(continue_prompt)
+                        except (EOFError, KeyboardInterrupt):
+                            self.lastcmd = ""
+                            print('\n')
+                            return
+                    else:
+                        self.stdout.write(continue_prompt)
+                        self.stdout.flush()
+                        line = self.stdin.readline()
+                        if not len(line):
+                            self.lastcmd = ""
+                            self.stdout.write('\n')
+                            self.stdout.flush()
+                            return
+                        else:
+                            line = line.rstrip('\r\n')
```

Comment:
The logic here seems to be at least somewhat based on similar logic in `cmd.Cmd.cmdloop`, with some subtle changes.

Is there any chance the logic could be shared a little better (maybe through another `cmd.Cmd` method), or are they just too different? If they're too different, maybe just add a comment explaining what the differences are, and why?

Reply:
The major issue here is the hooks, for example, `preloop()`. `pdb` implemented `preloop()` so if we call `self.cmdloop()` directly it will trigger the hook and print stuff that we don't want. Also `cmdloop` will actually parse the input and look for the according commands first. Do you suggest we create a new `cmd.Cmd` object to handle this? Not sure if the code would be cleaner. Taking advantage of the fact that `Pdb` is a sublclass of `cmd.Cmd` won't work I believe.